File: Programação orientada a objetos/POO 241 - Trab 1 - Pedro Quedevez/Regras.java
import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.text.ParseException;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Locale;

public class Regras {

    private LocalDate dataInicio;
    private LocalDate dataFim;
    private ArrayList<String> qualis1;
    private ArrayList<String> pontos;
    private int qtdeAnosPontos;
    private ArrayList<String> qualis2;
    private ArrayList<Integer> quantidadesMinimas;
    private int qtdeAnosArtigo;
    private Float pontuacaoMin;

    private static ArrayList<Regras> datasCadastradas = new ArrayList<Regras>();

    private static Boolean verificaDataValida(LocalDate inicio, LocalDate fim) {
        for (Regras regra : datasCadastradas) {
            if (inicio.isAfter(regra.dataInicio) && inicio.isBefore(regra.dataFim)
                    || fim.isAfter(regra.dataInicio) && fim.isBefore(regra.dataFim) || regra.dataInicio.equals(inicio)
                    || regra.dataFim.equals(fim)) {
                return false;
            }
            // agora para fazer as mesmas verificações na ordem inversa
            if (inicio.isBefore(regra.dataInicio) && fim.isAfter(regra.dataFim)
                    || inicio.isAfter(regra.dataInicio) && fim.isBefore(regra.dataFim)) {
                return false;
            }
        }
        return true;
    }

    public Regras(
            String csvLine) throws excecao {
        String[] campos = csvLine.split(";");
        try {
            if (campos.length == 9) {

                DecimalFormatSymbols symbols = new DecimalFormatSymbols(Locale.getDefault());
                symbols.setDecimalSeparator(',');
                DecimalFormat decimalFormat = new DecimalFormat("#.##", symbols);
                DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy");

                // this.dataInicio = new Date();

                this.dataInicio = LocalDate.parse(campos[0].trim(), formatter);
                this.dataFim = LocalDate.parse(campos[1].trim(), formatter);

                if (!verificaDataValida(dataInicio, dataFim)) {
                    throw new excecao("Múltiplas regras de pontuação para o mesmo período: "
                            + formatter.format(dataInicio) + " : "
                            + formatter.format(dataFim) + ".");
                }

                String[] qualis1 = campos[2].split("-");
                this.qualis1 = new ArrayList<String>();
                for (String quali : qualis1) {
                    if (!Arrays.asList(Qualificacoes.qualisValidos).contains(quali)) {
                        throw new excecao(
                                "Qualis desconhecido para regras de " + campos[0] + ": " + quali + ".");
                    }
                    this.qualis1.add(quali);
                }
                String[] pontos = campos[3].split("-");
                this.pontos = new ArrayList<String>();
                for (String ponto : pontos) {
                    this.pontos.add(ponto);
                }
                this.qtdeAnosPontos = Integer.parseInt(campos[4].trim());

                String[] qualis2 = campos[5].split("-");
                this.qualis2 = new ArrayList<String>();
                for (String quali : qualis2) {
                    if (!Arrays.asList(Qualificacoes.qualisValidos).contains(quali)) {
                        throw new excecao(
                                "Qualis desconhecido para regras de " + campos[0].trim() + ": " + quali + ".");
                    }
                    this.qualis2.add(quali);
                }

                String[] quantidadesMinimas = campos[6].split("-");
                this.quantidadesMinimas = new ArrayList<Integer>();
                for (String quantidade : quantidadesMinimas) {
                    int quant = Integer.parseInt(quantidade);
                    this.quantidadesMinimas.add(quant);
                }

                this.qtdeAnosArtigo = Integer.parseInt(campos[7]);
                try {
                    this.pontuacaoMin = decimalFormat.parse(campos[8].trim()).floatValue();
                } catch (Exception e) {
                    throw new excecao("Erro de formatacao");
                }
            }
        } catch (java.lang.NumberFormatException e) {
            throw new excecao("Erro de formatacao");
        }
        datasCadastradas.add(this);

    }

    public Float[] constroiVetorPontos() {

        DecimalFormatSymbols symbols = new DecimalFormatSymbols(Locale.getDefault());
        symbols.setDecimalSeparator(',');
        DecimalFormat decimalFormat = new DecimalFormat("#.##", symbols);

        String todosQualis[] = { "A1", "A2", "A3", "A4", "B1", "B2", "B3", "B4", "B5", "C" }; // vai usar i para iterar;
        Float vetorPontos[] = { 0.0f, 0.0f, 0.0f, 0.0f, 0.0f, 0.0f, 0.0f, 0.0f, 0.0f, 0.0f }; // vai usar i para iterar;
        ArrayList<String> limitesPontos = this.getQualis1(); // o valor do index vai alterar quando equals for
                                                             // verdadeiro;
        ArrayList<String> pontos = this.getPontos(); // o valor do index vai alterar quando equals for verdadeiro;
                                                     // possuem mesmo tamanho;
        int contAux = 0;
        Float pontuacaoAtual = 0.0f;

        for (int i = 0; i < vetorPontos.length; i++) {
            if (limitesPontos.get(contAux).equals(todosQualis[i])) { // verifica quando se atinge um dos limites para a
                                                                     // mudança de pontuação;

                try {
                    pontuacaoAtual = decimalFormat.parse(pontos.get(contAux)).floatValue(); // muda a pontuação a ser
                                                                                            // atribuída de acordo
                } catch (ParseException e) {
                    e.printStackTrace();
                }

                contAux++; // com os limites;
                if (contAux == limitesPontos.size()) {
                    contAux--; // gambiarra para não dar erro de index out of bounds;
                }
                // System.out.println(pontuacaoAtual + " " + contAux);
            }
            vetorPontos[i] = pontuacaoAtual; // armazena a pontuação no vetor;
        }
        return vetorPontos;
    }

    public static Regras getRegraAtual(ArrayList<Regras> regras, int ano) {
        for (Regras regra : regras) {
            LocalDate dataVerificacao = LocalDate.of(ano, 01, 01);
            if ((dataVerificacao.isAfter(regra.dataInicio) && dataVerificacao.isBefore(regra.dataFim))
                    || dataVerificacao.equals(regra.dataInicio) || dataVerificacao.equals(regra.dataFim)) {
                return regra;
            }
        }

        return null;
    }

    @Override
    public String toString() {
        return dataInicio + ";" + dataFim + ";" + qualis1 + ";" + pontos + ";" + qtdeAnosPontos + ";" + qualis2 + ";"
                + quantidadesMinimas + ";" + qtdeAnosArtigo + ";" + pontuacaoMin;
    }

    public LocalDate getDataInicio() {
        return dataInicio;
    }

    public void setDataInicio(LocalDate dataInicio) {
        this.dataInicio = dataInicio;
    }

    public LocalDate getDataFim() {
        return dataFim;
    }

    public void setDataFim(LocalDate dataFim) {
        this.dataFim = dataFim;
    }

    public ArrayList<String> getQualis1() {
        return qualis1;
    }

    public void setQualis1(ArrayList<String> qualis1) {
        this.qualis1 = qualis1;
    }

    public ArrayList<String> getPontos() {
        return pontos;
    }

    public void setPontos(ArrayList<String> pontos) {
        this.pontos = pontos;
    }

    public Integer getQtdeAnosPontos() {
        return qtdeAnosPontos;
    }

    public void setQtdeAnosPontos(Integer qtdeAnosPontos) {
        this.qtdeAnosPontos = qtdeAnosPontos;
    }

    public ArrayList<String> getQualis2() {
        return qualis2;
    }

    public void setQualis2(ArrayList<String> qualis2) {
        this.qualis2 = qualis2;
    }

    public ArrayList<Integer> getQuantidadesMinimas() {
        return quantidadesMinimas;
    }

    public void setQuantidadesMinimas(ArrayList<Integer> quantidadesMinimas) {
        this.quantidadesMinimas = quantidadesMinimas;
    }

    public Integer getQtdeAnosArtigo() {
        return qtdeAnosArtigo;
    }

    public void setQtdeAnosArtigo(Integer qtdeAnosArtigo) {
        this.qtdeAnosArtigo = qtdeAnosArtigo;
    }

    public Float getPontuacaoMin() {
        return pontuacaoMin;
    }

    public void setPontuacaoMin(Float pontuacaoMin) {
        this.pontuacaoMin = pontuacaoMin;
    }

}

File: Programação orientada a objetos/POO-241 - Trab 2 - Pedro_Quedevez/Regras.cpp
#include <iostream>
#include "Tokenizer.hpp"
#include "Regras.hpp"
#include "Excecao.hpp"
#include "ItemListaPublicacoes.hpp"

using namespace std;
using namespace cpp_util;

Regras::Regras()
{
}

Regras::Regras(string linhaCsv)
{
    Tokenizer tokens(linhaCsv, ';');
    vector<string> campos = tokens.remaining();
    try
    {

        this->setDataInicio(campos[0]);
        this->setDataFim(campos[1]);

        Tokenizer tokensQualis1(campos[2], '-');
        vector<string> qualis1 = tokensQualis1.remaining();
        for (string q : qualis1)
        {
            if (ordemQualisMap.find(q) == ordemQualisMap.end())
            {
                throw Excecao("Qualis desconhecido para regras de " + campos[0] + ": " + q + ".");
            }
        }
        this->setQualis1(qualis1);

        Tokenizer tokensPontos(campos[3], '-');
        vector<string> pontos = tokensPontos.remaining();
        this->setPontos(pontos);

        this->setQtdeAnosPontos(stoi(campos[4]));

        Tokenizer tokensQualis2(campos[5], '-');
        vector<string> qualis2 = tokensQualis2.remaining();
        this->setQualis2(qualis2);

        Tokenizer tokensQuantidadesMinimas(campos[6], '-');
        vector<string> quantidadesMinimas = tokensQuantidadesMinimas.remaining();
        vector<int> quantidadesMinimasInt;

        for (string s : quantidadesMinimas)
        {
            quantidadesMinimasInt.push_back(stoi(s));
        }

        this->setQuantidadesMinimas(quantidadesMinimasInt);
        this->setQtdeAnosArtigo(stoi(campos[7]));
        this->setPontuacaoMin(stof(campos[8]));
    }
    catch (Excecao e)
    {
        throw e;
    }
};

Regras Regras::getRegraAtual(vector<Regras> regras, int anoAtual)
{

    for (Regras r : regras)
    {
        Tokenizer tokensInicio(r.getDataInicio(), '/');
        vector<string> dataInicio = tokensInicio.remaining();
        Tokenizer tokensFim(r.getDataFim(), '/');
        vector<string> dataFim = tokensFim.remaining();

        if (anoAtual >= stoi(dataInicio[2]) && anoAtual <= stoi(dataFim[2]))
        {
            return r;
        }
    }

    Regras regrasNull = Regras();
    return regrasNull;
}

string Regras::getDataInicio()
{
    return this->dataInicio;
};

string Regras::getDataFim()
{
    return this->dataFim;
};

vector<string> Regras::getQualis1()
{
    return this->qualis1;
};

vector<string> Regras::getPontos()
{
    return this->pontos;
};

int Regras::getQtdeAnosPontos()
{
    return this->qtdeAnosPontos;
};

vector<string> Regras::getQualis2()
{
    return this->qualis2;
};

vector<int> Regras::getQuantidadesMinimas()
{
    return this->quantidadesMinimas;
};

int Regras::getQtdeAnosArtigo()
{
    return this->qtdeAnosArtigo;
};

float Regras::getPontuacaoMin()
{
    return this->pontuacaoMin;
};

void Regras::setDataInicio(string dataInicio)
{
    this->dataInicio = dataInicio;
};

void Regras::setDataFim(string dataFim)
{
    this->dataFim = dataFim;
};

void Regras::setQualis1(vector<string> qualis1)
{
    this->qualis1 = qualis1;
};

void Regras::setPontos(vector<string> pontos)
{
    this->pontos = pontos;
};

void Regras::setQtdeAnosPontos(int qtdeAnosPontos)
{
    this->qtdeAnosPontos = qtdeAnosPontos;
};

void Regras::setQualis2(vector<string> qualis2)
{
    this->qualis2 = qualis2;
};

void Regras::setQuantidadesMinimas(vector<int> quantidadesMinimas)
{
    this->quantidadesMinimas = quantidadesMinimas;
};

void Regras::setQtdeAnosArtigo(int qtdeAnosArtigo)
{
    this->qtdeAnosArtigo = qtdeAnosArtigo;
};

void Regras::setPontuacaoMin(float pontuacaoMin)
{
    this->pontuacaoMin = pontuacaoMin;
};

ostream &operator<<(ostream &r, Regras &regras)
{
    r << "Data Inicio: " << regras.getDataInicio() << endl;
    r << "Data Fim: " << regras.getDataFim() << endl;
    r << "Qualis1: ";
    for (string q : regras.getQualis1())
    {
        r << q << " ";
    }
    r << endl;
    r << "Pontos: ";
    for (string p : regras.getPontos())
    {
        r << p << " ";
    }
    r << endl;
    r << "Qtde Anos Pontos: " << regras.getQtdeAnosPontos() << endl;
    r << "Qualis2: ";
    for (string q : regras.getQualis2())
    {
        r << q << " ";
    }
    r << endl;
    r << "Quantidades Minimas: ";
    for (int q : regras.getQuantidadesMinimas())
    {
        r << q << " ";
    }
    r << endl;
    r << "Qtde Anos Artigo: " << regras.getQtdeAnosArtigo() << endl;
    r << "Pontuacao Min: " << regras.getPontuacaoMin() << endl;
    return r;
};